src/lint/concepts.nim

```diff
@@ -1,6 +1,81 @@
-import std/os
+import std/[json, os, strutils]
+import ".."/helpers
 import "."/validators
 
+proc isUrlLike(s: string): bool =
+  ## Returns true if `s` starts with `http://`, `https://` or `www`.
+  # For now, this is deliberately simplistic. We probably don't need
+  # sophisticated URL checking, and we don't want to use Nim's stdlib regular
+  # expressions because that would add a dependency on PCRE.
+  if s.startsWith("http"):
+    if s.continuesWith("://", 4) or s.continuesWith("s://", 4):
+      result = true
+  elif s.startsWith("www"):
+    result = true
+
+proc isValidLinkObject(data: JsonNode, context: string, path: string): bool =
+  ## Returns true if `data` is a `JObject` that satisfies all of the below:
+  ## - has a `url` key, with a value that is a URL-like string.
+  ## - has a `description` key, with a value that is a non-empty, non-blank string.
+  ## - if it has a `icon_url` key, the corresponding value is a URL-like string.
+  if isObject(data, context, path):
+    result = true
+
+    if checkString(data, "url", path):
+      let s = data["url"].getStr()
+      if not isUrlLike(s):
+        result.setFalseAndPrint("Not a valid URL: " & s, path)
+    else:
+      result = false
+
+    if not checkString(data, "description", path):
+      result = false
+
+    if data.hasKey("icon_url"):
```

Comment:
Do you need this call? Doesn't `checkString` gracefully handle optional values?

Reply:
> Do you need this call?

With the current `checkString`, unfortunately yes.

The full context is:
```Nim
    if data.hasKey("icon_url"):
      if checkString(data, "icon_url", path, isRequired = false):
        let s = data["icon_url"].getStr()
```

`icon_url` is optional, so we call `checkString` with `isRequired = false`. That means that `checkString` returns `true` if `icon_url` is missing, which means the `data["icon_url"].getStr()` would throw an exception if we don't have the first line.

It is indeed pretty ugly, and we're performing `hasKey` twice.

In the future I'm thinking about splitting `checkString` into `hasString` and `isString` (and the same for `checkBoolean`, and `checkInteger`), similar to the previous split of `checkArray` into `hasArray` and `isArray`.